Crash when setting `EXPOSE_OBJECT_API=true` with pachd 1.8.6

<!-- This form is for bug reports ONLY! 

If you're looking for help, please check:

Docs: http://docs.pachyderm.io/en/latest/
Slack: http://slack.pachyderm.io/

-->

**What happened?**:
I changed the value of the EXPOSE_OBJECT_API from 'false' to 'true' and pachd crashed on startup with this output:
```
2019-09-10T17:09:10Z WARNING TLS disabled: could not stat public cert at /pachd-tls-cert/tls.crt: stat /pachd-tls-cert/tls.crt: no such file or directory
2019-09-10T17:09:10Z WARNING TLS disabled: could not stat private key at /pachd-tls-cert/tls.key: stat /pachd-tls-cert/tls.key: no such file or directory
2019-09-10T17:09:10Z INFO Using cloudfront security credentials - keypair ID (APKAJKDBUNGUQRHTVH6Q) - to sign cloudfront URLs
2019-09-10T17:09:10Z INFO validating kubernetes access returned no errors
2019-09-10T17:09:10Z INFO Using cloudfront security credentials - keypair ID (APKAJKDBUNGUQRHTVH6Q) - to sign cloudfront URLs
2019-09-10T17:09:10Z INFO validating kubernetes access returned no errors
panic: duplicate registration of group object

goroutine 131 [running]:
github.com/pachyderm/pachyderm/src/server/vendor/github.com/golang/groupcache.newGroup(0x219acf8, 0x6, 0x0, 0x243b440, 0xc0002f0550, 0x0, 0x0, 0x0)
        src/github.com/pachyderm/pachyderm/src/server/vendor/github.com/golang/groupcache/groupcache.go:96 +0x338
github.com/pachyderm/pachyderm/src/server/vendor/github.com/golang/groupcache.NewGroup(0x219acf8, 0x6, 0x0, 0x243b440, 0xc0002f0550, 0x0)
        src/github.com/pachyderm/pachyderm/src/server/vendor/github.com/golang/groupcache/groupcache.go:84 +0x5b
github.com/pachyderm/pachyderm/src/server/pfs/server.newObjBlockAPIServer(0xc00005a04b, 0x4, 0x0, 0xc0001ca060, 0x19, 0x2475200, 0xc00077c8c0, 0x0, 0x0, 0x2260e98, ...)
        src/github.com/pachyderm/pachyderm/src/server/pfs/server/obj_block_api_server.go:97 +0x26e
github.com/pachyderm/pachyderm/src/server/pfs/server.newAmazonBlockAPIServer(0xc00005a04b, 0x4, 0x0, 0xc0001ca060, 0x19, 0x3a355a0, 0x2165f60, 0xc0003670e0)
        src/github.com/pachyderm/pachyderm/src/server/pfs/server/obj_block_api_server.go:177 +0xb1
github.com/pachyderm/pachyderm/src/server/pfs/server.NewBlockAPIServer(0xc00005a04a, 0x5, 0x0, 0xc0000560b0, 0x6, 0xc0001ca060, 0x19, 0xc000058075, 0x15, 0xc000058049, ...)
        src/github.com/pachyderm/pachyderm/src/server/pfs/server/server.go:52 +0x284
main.doFullMode.func7.1(0xc0004ed080, 0x6, 0x6)
        src/github.com/pachyderm/pachyderm/src/server/cmd/pachd/main.go:355 +0xd65
github.com/pachyderm/pachyderm/src/server/vendor/github.com/pachyderm/pachyderm/src/client/pkg/grpcutil.Serve(0xc0005604b0, 0x1, 0x1, 0x7f36ba658600, 0xc000001c80)
        src/github.com/pachyderm/pachyderm/src/server/vendor/github.com/pachyderm/pachyderm/src/client/pkg/grpcutil/server.go:107 +0x2ad
main.doFullMode.func7(0x8, 0x2265888)
        src/github.com/pachyderm/pachyderm/src/server/cmd/pachd/main.go:310 +0x24c
github.com/pachyderm/pachyderm/src/server/vendor/golang.org/x/sync/errgroup.(*Group).Go.func1(0xc000560450, 0xc00043e150)
        src/github.com/pachyderm/pachyderm/src/server/vendor/golang.org/x/sync/errgroup/errgroup.go:58 +0x57
created by github.com/pachyderm/pachyderm/src/server/vendor/golang.org/x/sync/errgroup.(*Group).Go
        src/github.com/pachyderm/pachyderm/src/server/vendor/golang.org/x/sync/errgroup/errgroup.go:55 +0x66
```

**What you expected to happen?**:
I expected the object API to be enabled so I could use `pachctl get-object`

**How to reproduce it (as minimally and precisely as possible)?**:

**Anything else we need to know?**:

**Environment?**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T11:13:54Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.7", GitCommit:"65ecaf0671341311ce6aea0edab46ee69f65d59e", GitTreeState:"clean", BuildDate:"2019-01-24T19:22:45Z", GoVersion:"go1.10.7", Compiler:"gc", Platform:"linux/amd64"}
```

- Pachyderm CLI and pachd server version (use `pachctl version`): 1.8.6 for both

- Cloud provider (e.g. aws, azure, gke) or local deployment (e.g. minikube vs dockerized k8s):
AWS, object store on S3

- OS (e.g. from /etc/os-release):
ubuntu 18.04

- Others:

--- a/src/server/auth/server/auth_test.go
+++ b/src/server/auth/server/auth_test.go
@@ -1451,6 +1451,67 @@ func TestStopAndDeletePipeline(t *testing.T) {
 	deleteAll(t)
 }
 
+// TestStopJob just confirms that the StopJob API works when auth is on
+func TestStopJob(t *testing.T) {
+	if testing.Short() {
+		t.Skip("Skipping integration tests in short mode")
+	}
+	deleteAll(t)
+	alice := tu.UniqueString("alice")
+	aliceClient := getPachClient(t, alice)
+
+	// alice creates a repo
+	repo := tu.UniqueString(t.Name())
+	require.NoError(t, aliceClient.CreateRepo(repo))
+	require.ElementsEqual(t, entries(alice, "owner"), getACL(t, aliceClient, repo))
+	_, err := aliceClient.PutFile(repo, "master", "/file", strings.NewReader("test"))
+	require.NoError(t, err)
+
+	// alice creates a pipeline
+	pipeline := tu.UniqueString("alice-pipeline")
+	require.NoError(t, aliceClient.CreatePipeline(
+		pipeline,
+		"", // default image: ubuntu:16.04
+		[]string{"bash"},
+		[]string{"sleep 600"},
+		&pps.ParallelismSpec{Constant: 1},
+		client.NewPFSInput(repo, "/*"),
+		"", // default output branch: master
+		false,
+	))
+	// Make sure the input and output repos have non-empty ACLs
+	require.ElementsEqual(t,
+		entries(alice, "owner", pl(pipeline), "reader"), getACL(t, aliceClient, repo))
+	require.ElementsEqual(t,
+		entries(alice, "owner", pl(pipeline), "writer"), getACL(t, aliceClient, pipeline))
+
+	// Stop the first job in 'pipeline'
+	var jobID string
+	require.NoErrorWithinTRetry(t, 30*time.Second, func() error {
+		jobs, err := aliceClient.ListJob(pipeline, nil /*inputs*/, nil /*output*/, -1 /*history*/, true /* full */)
+		if err != nil {
+			return err
+		}
+		if len(jobs) != 1 {
+			return fmt.Errorf("expected one job but got %d", len(jobs))
+		}
+		jobID = jobs[0].Job.ID
+		return nil
+	})
+
+	require.NoError(t, aliceClient.StopJob(jobID))
+	require.NoErrorWithinTRetry(t, 30*time.Second, func() error {
+		ji, err := aliceClient.InspectJob(jobID, false)
+		if err != nil {
+			return fmt.Errorf("could not inspect job %q: %v", jobID, err)
+		}
+		if ji.State != pps.JobState_JOB_KILLED {
+			return fmt.Errorf("expected job %q to be in JOB_KILLED but was in %s", jobID, ji.State.String())
+		}
+		return nil
+	})
+}
+
 // Test ListRepo checks that the auth information returned by ListRepo and
 // InspectRepo is correct.
 // TODO(msteffen): This should maybe go in pachyderm_test, since ListRepo isn't

--- a/src/server/cmd/pachd/main.go
+++ b/src/server/cmd/pachd/main.go
@@ -153,7 +153,7 @@ func doSidecarMode(config interface{}) (retErr error) {
 				if err != nil {
 					return fmt.Errorf("units.RAMInBytes: %v", err)
 				}
-				blockAPIServer, err := pfs_server.NewBlockAPIServer(env.StorageRoot, blockCacheBytes, env.StorageBackend, net.JoinHostPort(env.EtcdHost, env.EtcdPort))
+				blockAPIServer, err := pfs_server.NewBlockAPIServer(env.StorageRoot, blockCacheBytes, env.StorageBackend, net.JoinHostPort(env.EtcdHost, env.EtcdPort), false)
 				if err != nil {
 					return fmt.Errorf("pfs.NewBlockAPIServer: %v", err)
 				}
@@ -406,7 +406,8 @@ func doFullMode(config interface{}) (retErr error) {
 						blockAPIServer, err := pfs_server.NewBlockAPIServer(
 							env.StorageRoot,
 							0 /* = blockCacheBytes (disable cache) */, env.StorageBackend,
-							etcdAddress)
+							etcdAddress,
+							true /* duplicate */)
 						if err != nil {
 							return fmt.Errorf("pfs.NewBlockAPIServer: %v", err)
 						}
@@ -486,7 +487,7 @@ func doFullMode(config interface{}) (retErr error) {
 						return fmt.Errorf("units.RAMInBytes: %v", err)
 					}
 					blockAPIServer, err := pfs_server.NewBlockAPIServer(
-						env.StorageRoot, blockCacheBytes, env.StorageBackend, etcdAddress)
+						env.StorageRoot, blockCacheBytes, env.StorageBackend, etcdAddress, false)
 					if err != nil {
 						return fmt.Errorf("pfs.NewBlockAPIServer: %v", err)
 					}

--- a/src/server/pfs/server/obj_block_api_server.go
+++ b/src/server/pfs/server/obj_block_api_server.go
@@ -65,10 +65,21 @@ type objBlockAPIServer struct {
 	objectIndexesLock sync.RWMutex
 }
 
-// In test mode, we use unique names for cache groups, since we might want
-// to run multiple block servers locally, which would conflict if groups
-// had the same name. We also do not report stats to prometheus
-func newObjBlockAPIServer(dir string, cacheBytes int64, etcdAddress string, objClient obj.Client, test bool) (*objBlockAPIServer, error) {
+// newObjBlockAPIServer creates a new struct for handling Pachyderm Object API
+// reqquests.
+//
+// 'duplicate' indicates that this ObjBlockAPIServer is running alongside
+// another ObjBlockAPIServer and prevents collisions in their groupcache groups
+// and cache stats. Each process may have at most one ObjBlockAPIServer with
+// 'duplicate' set to false, but may have zero. Currently (2019-10-7) there are
+// two situations where a single process may run multiple ObjBlockAPIServers:
+// 1. Pachd, with EXPOSE_OBJECT_API=true (i.e. integration testing). In this
+//    case, the 'hidden' ObjBlockAPIServer will have duplicate=false and the
+//    'exposed' ObjBlockAPIServer (which would not exist in production) will
+//    have duplicate=true, and not use the cache or export cache stats
+// 2. PFS storage tests, which create several local ObjBlockAPIServers (none of
+//    which are primary but cannot collide)
+func newObjBlockAPIServer(dir string, cacheBytes int64, etcdAddress string, objClient obj.Client, duplicate bool) (*objBlockAPIServer, error) {
 	// defensive measure to make sure storage is working and error early if it's not
 	// this is where we'll find out if the credentials have been misconfigured
 	if err := obj.TestStorage(context.Background(), objClient); err != nil {
@@ -88,7 +99,7 @@ func newObjBlockAPIServer(dir string, cacheBytes int64, etcdAddress string, objC
 	objectInfoGroupName := "objectInfo"
 	blockGroupName := "block"
 
-	if test {
+	if duplicate {
 		uuid := uuid.New()
 		objectGroupName += uuid
 		tagGroupName += uuid
@@ -101,7 +112,7 @@ func newObjBlockAPIServer(dir string, cacheBytes int64, etcdAddress string, objC
 	s.objectInfoCache = groupcache.NewGroup(objectInfoGroupName, oneCacheShare*objectInfoCacheShares, groupcache.GetterFunc(s.objectInfoGetter))
 	s.blockCache = groupcache.NewGroup(blockGroupName, oneCacheShare*blockCacheShares, groupcache.GetterFunc(s.blockGetter))
 
-	if !test {
+	if !duplicate {
 		RegisterCacheStats("tag", &s.tagCache.Stats)
 		RegisterCacheStats("object", &s.objectCache.Stats)
 		RegisterCacheStats("object_info", &s.objectInfoCache.Stats)
@@ -165,44 +176,44 @@ func (s *objBlockAPIServer) getGeneration() int {
 	return s.generation
 }
 
-func newMinioBlockAPIServer(dir string, cacheBytes int64, etcdAddress string) (*objBlockAPIServer, error) {
+func newMinioBlockAPIServer(dir string, cacheBytes int64, etcdAddress string, duplicate bool) (*objBlockAPIServer, error) {
 	objClient, err := obj.NewMinioClientFromSecret("")
 	if err != nil {
 		return nil, err
 	}
-	return newObjBlockAPIServer(dir, cacheBytes, etcdAddress, objClient, false)
+	return newObjBlockAPIServer(dir, cacheBytes, etcdAddress, objClient, duplicate)
 }
 
-func newAmazonBlockAPIServer(dir string, cacheBytes int64, etcdAddress string) (*objBlockAPIServer, error) {
+func newAmazonBlockAPIServer(dir string, cacheBytes int64, etcdAddress string, duplicate bool) (*objBlockAPIServer, error) {
 	objClient, err := obj.NewAmazonClientFromSecret("")
 	if err != nil {
 		return nil, err
 	}
-	return newObjBlockAPIServer(dir, cacheBytes, etcdAddress, objClient, false)
+	return newObjBlockAPIServer(dir, cacheBytes, etcdAddress, objClient, duplicate)
 }
 
-func newGoogleBlockAPIServer(dir string, cacheBytes int64, etcdAddress string) (*objBlockAPIServer, error) {
+func newGoogleBlockAPIServer(dir string, cacheBytes int64, etcdAddress string, duplicate bool) (*objBlockAPIServer, error) {
 	objClient, err := obj.NewGoogleClientFromSecret("")
 	if err != nil {
 		return nil, err
 	}
-	return newObjBlockAPIServer(dir, cacheBytes, etcdAddress, objClient, false)
+	return newObjBlockAPIServer(dir, cacheBytes, etcdAddress, objClient, duplicate)
 }
 
-func newMicrosoftBlockAPIServer(dir string, cacheBytes int64, etcdAddress string) (*objBlockAPIServer, error) {
+func newMicrosoftBlockAPIServer(dir string, cacheBytes int64, etcdAddress string, duplicate bool) (*objBlockAPIServer, error) {
 	objClient, err := obj.NewMicrosoftClientFromSecret("")
 	if err != nil {
 		return nil, err
 	}
-	return newObjBlockAPIServer(dir, cacheBytes, etcdAddress, objClient, false)
+	return newObjBlockAPIServer(dir, cacheBytes, etcdAddress, objClient, duplicate)
 }
 
-func newLocalBlockAPIServer(dir string, cacheBytes int64, etcdAddress string) (*objBlockAPIServer, error) {
+func newLocalBlockAPIServer(dir string, cacheBytes int64, etcdAddress string, duplicate bool) (*objBlockAPIServer, error) {
 	objClient, err := obj.NewLocalClient(dir)
 	if err != nil {
 		return nil, err
 	}
-	return newObjBlockAPIServer(dir, cacheBytes, etcdAddress, objClient, true)
+	return newObjBlockAPIServer(dir, cacheBytes, etcdAddress, objClient, duplicate)
 }
 
 func (s *objBlockAPIServer) PutObject(server pfsclient.ObjectAPI_PutObjectServer) (retErr error) {

--- a/src/server/pfs/server/server.go
+++ b/src/server/pfs/server/server.go
@@ -41,14 +41,16 @@ func NewAPIServer(
 
 // NewBlockAPIServer creates a BlockAPIServer using the credentials it finds in
 // the environment
-func NewBlockAPIServer(dir string, cacheBytes int64, backend string, etcdAddress string) (BlockAPIServer, error) {
+// TODO(msteffen) accept serviceenv.ServiceEnv instead of 'dir', 'backend', and
+// 'duplicate'?
+func NewBlockAPIServer(dir string, cacheBytes int64, backend string, etcdAddress string, duplicate bool) (BlockAPIServer, error) {
 	switch backend {
 	case MinioBackendEnvVar:
 		// S3 compatible doesn't like leading slashes
 		if len(dir) > 0 && dir[0] == '/' {
 			dir = dir[1:]
 		}
-		blockAPIServer, err := newMinioBlockAPIServer(dir, cacheBytes, etcdAddress)
+		blockAPIServer, err := newMinioBlockAPIServer(dir, cacheBytes, etcdAddress, duplicate)
 		if err != nil {
 			return nil, err
 		}
@@ -58,20 +60,20 @@ func NewBlockAPIServer(dir string, cacheBytes int64, backend string, etcdAddress
 		if len(dir) > 0 && dir[0] == '/' {
 			dir = dir[1:]
 		}
-		blockAPIServer, err := newAmazonBlockAPIServer(dir, cacheBytes, etcdAddress)
+		blockAPIServer, err := newAmazonBlockAPIServer(dir, cacheBytes, etcdAddress, duplicate)
 		if err != nil {
 			return nil, err
 		}
 		return blockAPIServer, nil
 	case GoogleBackendEnvVar:
 		// TODO figure out if google likes leading slashses
-		blockAPIServer, err := newGoogleBlockAPIServer(dir, cacheBytes, etcdAddress)
+		blockAPIServer, err := newGoogleBlockAPIServer(dir, cacheBytes, etcdAddress, duplicate)
 		if err != nil {
 			return nil, err
 		}
 		return blockAPIServer, nil
 	case MicrosoftBackendEnvVar:
-		blockAPIServer, err := newMicrosoftBlockAPIServer(dir, cacheBytes, etcdAddress)
+		blockAPIServer, err := newMicrosoftBlockAPIServer(dir, cacheBytes, etcdAddress, duplicate)
 		if err != nil {
 			return nil, err
 		}
@@ -79,7 +81,7 @@ func NewBlockAPIServer(dir string, cacheBytes int64, backend string, etcdAddress
 	case LocalBackendEnvVar:
 		fallthrough
 	default:
-		blockAPIServer, err := newLocalBlockAPIServer(dir, cacheBytes, etcdAddress)
+		blockAPIServer, err := newLocalBlockAPIServer(dir, cacheBytes, etcdAddress, duplicate)
 		if err != nil {
 			return nil, err
 		}

--- a/src/server/pfs/server/testing.go
+++ b/src/server/pfs/server/testing.go
@@ -109,7 +109,11 @@ func GetPachClient(t testing.TB) *client.APIClient {
 	config.EtcdPort = etcdPort
 	config.PeerPort = uint16(pfsPort)
 	env := serviceenv.InitServiceEnv(config)
-	blockAPIServer, err := newLocalBlockAPIServer(root, localBlockServerCacheBytes, net.JoinHostPort(etcdHost, etcdPort))
+	blockAPIServer, err := newLocalBlockAPIServer(
+		root,
+		localBlockServerCacheBytes,
+		net.JoinHostPort(etcdHost, etcdPort),
+		true /* duplicate--see comment in newObjBlockAPIServer */)
 	require.NoError(t, err)
 	etcdPrefix := generateRandomString(32)
 	treeCache, err := hashtree.NewCache(testingTreeCacheSize)

--- a/src/server/pps/server/api_server.go
+++ b/src/server/pps/server/api_server.go
@@ -373,12 +373,13 @@ func (a *apiServer) validateKube() {
 	}
 }
 
-func checkLoggedIn(pachClient *client.APIClient) error {
-	_, err := pachClient.WhoAmI(pachClient.Ctx(), &auth.WhoAmIRequest{})
+func checkLoggedIn(pachClient *client.APIClient) (context.Context, error) {
+	ctx := pachClient.Ctx() // pachClient propagates auth info
+	_, err := pachClient.WhoAmI(ctx, &auth.WhoAmIRequest{})
 	if err != nil && !auth.IsErrNotActivated(err) {
-		return err
+		return nil, err
 	}
-	return nil
+	return ctx, nil
 }
 
 // authorizing a pipeline operation varies slightly depending on whether the
@@ -503,8 +504,8 @@ func (a *apiServer) CreateJob(ctx context.Context, request *pps.CreateJobRequest
 	func() { a.Log(request, nil, nil, 0) }()
 	defer func(start time.Time) { a.Log(request, response, retErr, time.Since(start)) }(time.Now())
 	pachClient := a.env.GetPachClient(ctx)
-	ctx = pachClient.Ctx() // pachClient will propagate auth info
-	if err := checkLoggedIn(pachClient); err != nil {
+	ctx, err := checkLoggedIn(pachClient)
+	if err != nil {
 		return nil, err
 	}
 
@@ -512,7 +513,7 @@ func (a *apiServer) CreateJob(ctx context.Context, request *pps.CreateJobRequest
 	if request.Stats == nil {
 		request.Stats = &pps.ProcessStats{}
 	}
-	_, err := col.NewSTM(ctx, a.env.GetEtcdClient(), func(stm col.STM) error {
+	_, err = col.NewSTM(ctx, a.env.GetEtcdClient(), func(stm col.STM) error {
 		jobPtr := &pps.EtcdJobInfo{
 			Job:           job,
 			OutputCommit:  request.OutputCommit,
@@ -541,7 +542,8 @@ func (a *apiServer) InspectJob(ctx context.Context, request *pps.InspectJobReque
 	func() { a.Log(request, nil, nil, 0) }()
 	defer func(start time.Time) { a.Log(request, response, retErr, time.Since(start)) }(time.Now())
 	pachClient := a.env.GetPachClient(ctx)
-	if err := checkLoggedIn(pachClient); err != nil {
+	ctx, err := checkLoggedIn(pachClient)
+	if err != nil {
 		return nil, err
 	}
 	if request.Job == nil && request.OutputCommit == nil {
@@ -824,6 +826,7 @@ func (a *apiServer) ListJob(ctx context.Context, request *pps.ListJobRequest) (r
 		}
 	}(time.Now())
 	pachClient := a.env.GetPachClient(ctx)
+	ctx = pachClient.Ctx() // pachClient will propagate auth info
 	var jobInfos []*pps.JobInfo
 	if err := a.listJob(pachClient, request.Pipeline, request.OutputCommit, request.InputCommit, request.History, request.Full, func(ji *pps.JobInfo) error {
 		jobInfos = append(jobInfos, ji)
@@ -859,7 +862,8 @@ func (a *apiServer) FlushJob(request *pps.FlushJobRequest, resp pps.API_FlushJob
 		a.Log(request, fmt.Sprintf("stream containing %d JobInfos", sent), retErr, time.Since(start))
 	}(time.Now())
 	pachClient := a.env.GetPachClient(resp.Context())
-	if err := checkLoggedIn(pachClient); err != nil {
+	ctx, err := checkLoggedIn(pachClient)
+	if err != nil {
 		return err
 	}
 	var toRepos []*pfs.Repo
@@ -889,7 +893,7 @@ func (a *apiServer) FlushJob(request *pps.FlushJobRequest, resp pps.API_FlushJob
 		}
 		// Even though the commit has been finished the job isn't necessarily
 		// finished yet, so we block on its state as well.
-		ji, err := a.InspectJob(resp.Context(), &pps.InspectJobRequest{Job: jis[0].Job, BlockState: true})
+		ji, err := a.InspectJob(ctx, &pps.InspectJobRequest{Job: jis[0].Job, BlockState: true})
 		if err != nil {
 			return err
 		}
@@ -902,11 +906,13 @@ func (a *apiServer) DeleteJob(ctx context.Context, request *pps.DeleteJobRequest
 	func() { a.Log(request, nil, nil, 0) }()
 	defer func(start time.Time) { a.Log(request, response, retErr, time.Since(start)) }(time.Now())
 	pachClient := a.env.GetPachClient(ctx)
-	if err := checkLoggedIn(pachClient); err != nil {
+	ctx = pachClient.Ctx() // pachClient will propagate auth info
+	ctx, err := checkLoggedIn(pachClient)
+	if err != nil {
 		return nil, err
 	}
 
-	_, err := col.NewSTM(ctx, a.env.GetEtcdClient(), func(stm col.STM) error {
+	_, err = col.NewSTM(ctx, a.env.GetEtcdClient(), func(stm col.STM) error {
 		return a.jobs.ReadWrite(stm).Delete(request.Job.ID)
 	})
 	if err != nil {
@@ -920,9 +926,11 @@ func (a *apiServer) StopJob(ctx context.Context, request *pps.StopJobRequest) (r
 	func() { a.Log(request, nil, nil, 0) }()
 	defer func(start time.Time) { a.Log(request, response, retErr, time.Since(start)) }(time.Now())
 	pachClient := a.env.GetPachClient(ctx)
-	if err := checkLoggedIn(pachClient); err != nil {
+	ctx, err := checkLoggedIn(pachClient)
+	if err != nil {
 		return nil, err
 	}
+	ctx = pachClient.Ctx() // pachClient will propagate auth info
 
 	// Lookup jobInfo
 	jobPtr := &pps.EtcdJobInfo{}
@@ -946,9 +954,11 @@ func (a *apiServer) RestartDatum(ctx context.Context, request *pps.RestartDatumR
 	func() { a.Log(request, nil, nil, 0) }()
 	defer func(start time.Time) { a.Log(request, response, retErr, time.Since(start)) }(time.Now())
 	pachClient := a.env.GetPachClient(ctx)
-	if err := checkLoggedIn(pachClient); err != nil {
+	ctx, err := checkLoggedIn(pachClient)
+	if err != nil {
 		return nil, err
 	}
+	ctx = pachClient.Ctx() // pachClient will propagate auth info
 
 	jobInfo, err := a.InspectJob(ctx, &pps.InspectJobRequest{
 		Job: request.Job,
@@ -967,7 +977,7 @@ func (a *apiServer) RestartDatum(ctx context.Context, request *pps.RestartDatumR
 // between ListDatum and ListDatumStream. When ListDatum is removed, this should
 // be inlined into ListDatumStream
 func (a *apiServer) listDatum(pachClient *client.APIClient, job *pps.Job, page, pageSize int64) (response *pps.ListDatumResponse, retErr error) {
-	if err := checkLoggedIn(pachClient); err != nil {
+	if _, err := checkLoggedIn(pachClient); err != nil {
 		return nil, err
 	}
 	response = &pps.ListDatumResponse{}
@@ -1145,8 +1155,7 @@ func (a *apiServer) ListDatum(ctx context.Context, request *pps.ListDatumRequest
 			a.Log(request, response, retErr, time.Since(start))
 		}
 	}(time.Now())
-	pachClient := a.env.GetPachClient(ctx)
-	return a.listDatum(pachClient, request.Job, request.Page, request.PageSize)
+	return a.listDatum(a.env.GetPachClient(ctx), request.Job, request.Page, request.PageSize)
 }
 
 // ListDatumStream implements the protobuf pps.ListDatumStream RPC
@@ -1156,8 +1165,7 @@ func (a *apiServer) ListDatumStream(req *pps.ListDatumRequest, resp pps.API_List
 	defer func(start time.Time) {
 		a.Log(req, fmt.Sprintf("stream containing %d DatumInfos", sent), retErr, time.Since(start))
 	}(time.Now())
-	pachClient := a.env.GetPachClient(resp.Context())
-	ldr, err := a.listDatum(pachClient, req.Job, req.Page, req.PageSize)
+	ldr, err := a.listDatum(a.env.GetPachClient(resp.Context()), req.Job, req.Page, req.PageSize)
 	if err != nil {
 		return err
 	}
@@ -1252,7 +1260,8 @@ func (a *apiServer) InspectDatum(ctx context.Context, request *pps.InspectDatumR
 	func() { a.Log(request, nil, nil, 0) }()
 	defer func(start time.Time) { a.Log(request, response, retErr, time.Since(start)) }(time.Now())
 	pachClient := a.env.GetPachClient(ctx)
-	if err := checkLoggedIn(pachClient); err != nil {
+	ctx, err := checkLoggedIn(pachClient)
+	if err != nil {
 		return nil, err
 	}
 	ctx = pachClient.Ctx() // pachClient will propagate auth info
@@ -2236,7 +2245,7 @@ func (a *apiServer) InspectPipeline(ctx context.Context, request *pps.InspectPip
 // Many functions (GetLogs, ListPipeline, CreateJob) need to inspect a pipeline,
 // so they call this instead of making an RPC
 func (a *apiServer) inspectPipeline(pachClient *client.APIClient, name string) (*pps.PipelineInfo, error) {
-	if err := checkLoggedIn(pachClient); err != nil {
+	if _, err := checkLoggedIn(pachClient); err != nil {
 		return nil, err
 	}
 	kubeClient := a.env.GetKubeClient()
@@ -2314,7 +2323,8 @@ func (a *apiServer) ListPipeline(ctx context.Context, request *pps.ListPipelineR
 		}
 	}(time.Now())
 	pachClient := a.env.GetPachClient(ctx)
-	if err := checkLoggedIn(pachClient); err != nil {
+	ctx, err := checkLoggedIn(pachClient)
+	if err != nil {
 		return nil, err
 	}
 	pipelineInfos := &pps.PipelineInfos{}
@@ -2385,7 +2395,8 @@ func (a *apiServer) DeletePipeline(ctx context.Context, request *pps.DeletePipel
 	func() { a.Log(request, nil, nil, 0) }()
 	defer func(start time.Time) { a.Log(request, response, retErr, time.Since(start)) }(time.Now())
 	pachClient := a.env.GetPachClient(ctx)
-	if err := checkLoggedIn(pachClient); err != nil {
+	ctx, err := checkLoggedIn(pachClient)
+	if err != nil {
 		return nil, err
 	}
 
@@ -2916,7 +2927,8 @@ func (a *apiServer) GarbageCollect(ctx context.Context, request *pps.GarbageColl
 	func() { a.Log(request, nil, nil, 0) }()
 	defer func(start time.Time) { a.Log(request, response, retErr, time.Since(start)) }(time.Now())
 	pachClient := a.env.GetPachClient(ctx)
-	if err := checkLoggedIn(pachClient); err != nil {
+	ctx, err := checkLoggedIn(pachClient)
+	if err != nil {
 		return nil, err
 	}
 	pipelineInfos, err := a.ListPipeline(ctx, &pps.ListPipelineRequest{})
@@ -3033,8 +3045,8 @@ func (a *apiServer) ActivateAuth(ctx context.Context, req *pps.ActivateAuthReque
 	func() { a.Log(req, nil, nil, 0) }()
 	defer func(start time.Time) { a.Log(req, resp, retErr, time.Since(start)) }(time.Now())
 	pachClient := a.env.GetPachClient(ctx)
-	ctx = pachClient.Ctx() // pachClient will propagate auth infothis list
-	if err := checkLoggedIn(pachClient); err != nil {
+	ctx, err := checkLoggedIn(pachClient)
+	if err != nil {
 		return nil, err
 	}
 

--- a/src/server/transaction/server/server_test.go
+++ b/src/server/transaction/server/server_test.go
@@ -108,7 +108,7 @@ func GetPachClient(t testing.TB) *client.APIClient {
 	config.PeerPort = uint16(testPort)
 	env := serviceenv.InitServiceEnv(config)
 
-	pfsBlockServer, err := pfsserver.NewBlockAPIServer(root, localBlockServerCacheBytes, pfsserver.LocalBackendEnvVar, net.JoinHostPort(etcdHost, etcdPort))
+	pfsBlockServer, err := pfsserver.NewBlockAPIServer(root, localBlockServerCacheBytes, pfsserver.LocalBackendEnvVar, net.JoinHostPort(etcdHost, etcdPort), true /* duplicate */)
 	require.NoError(t, err)
 
 	etcdPrefix := generateRandomString(32)

